main/src/ca/uwaterloo/flix/api/lsp/provider/Visitor.scala

```diff
@@ -0,0 +1,850 @@
+    if (!a.accept(tc.loc)) { return }
+
+    c.consumeTraitConstraint(tc)
+    visitTraitConstraintHead(tc.head)
+    visitType(tc.arg)
+  }
+
+  private def visitTraitConstraintHead(tcHead: TraitConstraint.Head)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(tcHead.loc)) { return }
+    c.consumeTraitConstraintHead(tcHead)
+  }
+
+  private def visitAssocTypeDef(tdefn: AssocTypeDef)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(tdefn.loc)) { return }
+    c.consumeAssocTypeDef(tdefn)
+    visitAssocTypeSymUse(tdefn.sym)
+    visitType(tdefn.arg)
+  }
+
+  private def visitAssocTypeSymUse(symUse: AssocTypeSymUse)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(symUse.loc)) { return }
+    c.consumeAssocTypeSymUse(symUse)
+  }
+
+  private def visitSig(sig: Sig)(implicit a: Acceptor, c: Consumer): Unit = {
+    // TODO fix hack. This hack is necessary because a signature currently does not have a location.
+    val insideSig = a.accept(sig.spec.loc) || sig.exp.exists(e => a.accept(e.loc))
+    if (!insideSig) { return }
+
+    c.consumeSig(sig)
+    sig.exp.foreach(visitExpr)
+  }
+
+  private def visitStruct(struct: Struct)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(struct.loc)) { return }
+    c.consumeStruct(struct)
+
+    visitAnnotations(struct.ann)
+    struct.tparams.foreach(visitTypeParam)
+    // TODO visit scheme?
+    struct.fields.map(_._2).foreach(visitStructField)
+  }
+
+  private def visitStructField(field: StructField)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(field.loc)) { return }
+    c.consumeStructField(field)
+    visitType(field.tpe)
+  }
+
+  private def visitTrait(t: Trait)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(t.loc)) { return }
+
+    c.consumeTrait(t)
+
+    visitAnnotations(t.ann)
+    visitTypeParam(t.tparam)
+    t.superTraits.foreach(visitTraitConstraint)
+    t.assocs.foreach(visitAssocTypeSig)
+    t.sigs.foreach(visitSig)
+    t.laws.foreach(visitDef)
+  }
+
+  private def visitAssocTypeSig(assoc: AssocTypeSig)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(assoc.loc)) { return }
+
+    c.consumeAssocTypeSig(assoc)
+
+    visitTypeParam(assoc.tparam)
+    visitKind(assoc.kind)
+    assoc.tpe.foreach(visitType)
+    
+  }
+
+  private def visitKind(kind: Kind)(implicit a: Acceptor, c: Consumer): Unit = {
+    // TODO it doesn't have a location, so does this even make sense?
+  }
+  
+  private def visitEffect(eff: Effect)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!(a.accept(eff.loc))) { return }
+    c.consumeEff(eff)
+
+    visitAnnotations(eff.ann)
+    eff.ops.foreach(visitOp)
+  }
+
+  private def visitOp(op: Op)(implicit a: Acceptor, c: Consumer): Unit = {
+    // TODO: hack that should eventually be fixed. Really it should be `op.loc`, but since op decls don't have locations, this hack is necessary for now
```

Comment:
Same comment as above.

Reply:
We ended up agreeing that this should infact be visited, right? Just like sigs and defs